src/main/java/net/hydromatic/lookml/LookmlWriter.java

```diff
@@ -0,0 +1,158 @@
+ * license agreements.  See the NOTICE file distributed with this
+ * work for additional information regarding copyright ownership.
+ * The LookML Authors license this file to you under the Apache
+ * License, Version 2.0 (the "License"); you may not use this
+ * file except in compliance with the License.  You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied.  See the License for the specific
+ * language governing permissions and limitations under the
+ * License.
+ */
+package net.hydromatic.lookml;
+
+import net.hydromatic.lookml.util.PairList;
+
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/** Writes fragments of LookML to a {@link StringBuilder},
+ * and keeps track of the current indentation level.*/
+class LookmlWriter {
+  /**
+   * Regex pattern for an identifier that does not need to be quoted.
+   */
+  static final Pattern SIMPLE_IDENTIFIER_PATTERN =
+      Pattern.compile("[a-zA-Z0-9_]*");
+
+  private final boolean pretty;
+  private int indent = 0;
+  private final int offset;
+  final StringBuilder buf;
+
+  LookmlWriter(boolean pretty, StringBuilder buf, int offset) {
+    this.pretty = pretty;
+    this.buf = buf;
+    this.offset = offset;
+  }
+
+  /**
+   * Adds a string value.
+   */
+  void string(String s) {
+    buf.append('"').append(s.replace("\"", "\\s")).append('"');
+  }
+
+  /**
+   * Adds a numeric value.
+   */
+  void number(Number number) {
+    buf.append(number);
+  }
+
+  /**
+   * Adds a code value.
+   */
+  void code(String s) {
+    buf.append(s).append(";;");
+  }
+
+  /**
+   * Adds an object.
+   */
+  void obj(PairList<String, Value> properties) {
+    if (properties.isEmpty()) {
+      buf.append("{}");
+    } else {
+      buf.append('{');
+      if (pretty) {
+        buf.append('\n');
+        indent += offset;
+        Spaces.append(buf, indent);
+      }
+      propertyList(properties);
+      if (pretty) {
+        buf.append('\n');
+        indent -= offset;
+        Spaces.append(buf, indent);
+      }
+      buf.append('}');
+    }
+  }
+
+  /**
+   * Adds a property list (an object without braces).
+   */
+  void propertyList(PairList<String, Value> properties) {
+    properties.forEachIndexed((i, property, value) -> {
+      if (i > 0) {
+        if (pretty) {
+          buf.append('\n');
+          Spaces.append(buf, indent);
+        } else {
+          buf.append(", ");
+        }
+      }
+      value.write(property, this);
+    });
+  }
+
+  void label(String name) {
+    identifier(name);
+    buf.append(pretty ? ": " : ":");
+  }
+
+  void identifier(String id) {
+    Matcher m = SIMPLE_IDENTIFIER_PATTERN.matcher(id);
```

Comment:
The events do not necessarily come from a parser. They might, for example, come from a program. So, someone could supply "`My table`", and the writer needs to decide whether this is valid, and if so, how to represent it in the syntax.

My assumption is that a compiled regex that checks for alphanumeric strings is pretty efficient. If that assumption is wrong it will show up during profiling and we can replace with custom code.